src/utils/pricing.ts

```diff
@@ -0,0 +1,119 @@
+  '0x956f47f50a910163d8bf957cf5846d573e7f87ca',
+  '0x4dd28568d05f09b02220b09c2cb307bfd837cb95'
+]
+
+let MINIMUM_ETH_LOCKED = BigDecimal.fromString('52')
+
+let Q192 = 2 ** 192
+export function sqrtPriceX96ToTokenPrices(sqrtPriceX96: BigInt, token0: Token, token1: Token): BigDecimal[] {
+  let num = sqrtPriceX96.times(sqrtPriceX96).toBigDecimal()
+  let denom = BigDecimal.fromString(Q192.toString())
+  let price1 = num
+    .div(denom)
+    .times(exponentToBigDecimal(BigInt.fromI32(token0.decimals)))
+    .div(exponentToBigDecimal(BigInt.fromI32(token1.decimals)))
+
+  let price0 = safeDiv(BigDecimal.fromString('1'), price1)
+  return [price0, price1]
+}
+
+export function getEthPriceInUSD(): BigDecimal {
+  // fetch eth prices for each stablecoin
+  let usdcPool = UniswapPool.load(USDC_WETH_03_POOL) // dai is token0
+  if (usdcPool !== null) {
+    return usdcPool.token0Price
+  } else {
+    return ZERO_BD
+  }
+}
+
+/**
+ * Search through graph to find derived Eth per token.
+ * @todo update to be derived ETH (add stablecoin estimates)
+ **/
+export function findEthPerToken(token: Token): BigDecimal {
+  if (token.id == WETH_ADDRESS) {
+    return ONE_BD
+  }
+  let whiteList = token.allowedPools
+  // for now just take USD from pool with greatest TVL
+  // need to update this to actually detect best rate based on liquidity distribution
+  let largestLiquidityETH = ZERO_BD
+  let priceSoFar = ZERO_BD
+  let bundle = Bundle.load('1')
+
+  // hardcoded fix for incorrect rates
+  // if allowed includes token - get the safe price
+  if (STABLE_COINS.includes(token.id)) {
+    priceSoFar = safeDiv(ONE_BD, bundle.ethPriceUSD)
+  } else {
+    for (let i = 0; i < whiteList.length; ++i) {
+      let poolAddress = whiteList[i]
+      let pool = UniswapPool.load(poolAddress)
+
+      if (pool.liquidity.gt(ZERO_BI)) {
+        if (pool.token0 == token.id) {
+          // allowed token is token1
+          let token1 = Token.load(pool.token1)
+          // get the derived ETH in pool
+          let ethLocked = pool.totalValueLockedToken1.times(token1.derivedETH)
+          if (ethLocked.gt(largestLiquidityETH) && ethLocked.gt(MINIMUM_ETH_LOCKED)) {
+            largestLiquidityETH = ethLocked
+            // token1 per our token * Eth per token1
+            priceSoFar = pool.token1Price.times(token1.derivedETH as BigDecimal)
+          }
+        }
+        if (pool.token1 == token.id) {
+          let token0 = Token.load(pool.token0)
+          // get the derived ETH in pool
+          let ethLocked = pool.totalValueLockedToken0.times(token0.derivedETH)
+          if (ethLocked.gt(largestLiquidityETH) && ethLocked.gt(MINIMUM_ETH_LOCKED)) {
+            largestLiquidityETH = ethLocked
+            // token0 per our token * ETH per token0
+            priceSoFar = pool.token0Price.times(token0.derivedETH as BigDecimal)
+          }
+        }
+      }
+    }
+  }
+  return priceSoFar // nothing was found return 0
```

Comment:
If I got this loop right, it'll go over every pool this token can use for USD estimations.
I don't expect to have many, but a minor optimization is to return early inside the loop when we update `priceSoFar`.

As far I can see, the values are never added up, only overwritten if any.

Did I read this right?

Reply:
Leandro, I think u cannot end early. 
U need to use this `if (ethLocked.gt(largestLiquidityETH)` and `largestLiquidityETH = ethLocked`

so the pool with more TVL decides the price

Reply:
Thanks for the explanation @anxolin ! That's it!